src/stacked_borrows/stack.rs

```diff
@@ -0,0 +1,258 @@
+    cache: StackCache,
+    /// On a read, we need to disable all `Unique` above the granting item. We can avoid most of
+    /// this scan by keeping track of the region of the borrow stack that may contain `Unique`s.
+    unique_range: Range<usize>,
+}
+
+/// A very small cache of searches of the borrow stack
+/// This maps tags to locations in the borrow stack. Any use of this still needs to do a
+/// probably-cold random access into the borrow stack to figure out what `Permission` an
+/// `SbTag` grants. We could avoid this by also storing the `Permission` in the cache, but
+/// most lookups into the cache are immediately followed by access of the full borrow stack anyway.
+#[derive(Clone, Debug)]
+struct StackCache {
+    tags: [SbTag; CACHE_LEN], // Hot in find_granting
+    idx: [usize; CACHE_LEN],  // Hot in grant
+}
+
+impl StackCache {
+    fn add(&mut self, idx: usize, tag: SbTag) {
+        self.tags.copy_within(0..CACHE_LEN - 1, 1);
+        self.tags[0] = tag;
+        self.idx.copy_within(0..CACHE_LEN - 1, 1);
+        self.idx[0] = idx;
+    }
+}
+
+impl PartialEq for Stack {
+    fn eq(&self, other: &Self) -> bool {
+        // All the semantics of Stack are in self.borrows, everything else is caching
+        self.borrows == other.borrows
+    }
+}
+
+impl Eq for Stack {}
+
+impl Stack {
+    /// Panics if any of the caching mechanisms have broken,
+    /// - The StackCache indices don't refer to the parallel tags,
+    /// - There are no Unique tags outside of first_unique..last_unique
+    #[cfg(feature = "expensive-debug-assertions")]
+    fn verify_cache_consistency(&self) {
+        for (tag, stack_idx) in self.cache.tags.iter().zip(self.cache.idx.iter()) {
+            assert_eq!(self.borrows[*stack_idx].tag, *tag);
+        }
+
+        for (idx, item) in self.borrows.iter().enumerate() {
+            if item.perm == Permission::Unique {
+                assert!(
+                    self.unique_range.contains(&idx),
+                    "{:?} {:?}",
+                    self.unique_range,
+                    self.borrows
+                );
+            }
+        }
+    }
+
+    /// Find the item granting the given kind of access to the given tag, and return where
+    /// it is on the stack.
+    pub fn find_granting(&mut self, access: AccessKind, tag: SbTag) -> Option<usize> {
+        #[cfg(feature = "expensive-debug-assertions")]
+        self.verify_cache_consistency();
+
+        match tag {
+            SbTag::Untagged =>
+                self.borrows
+                    .iter()
+                    .enumerate()
+                    .rev()
+                    .filter(|(_, item)| item.tag == SbTag::Untagged)
+                    .find_map(|(idx, item)| if item.perm.grants(access) { Some(idx) } else { None }),
+            SbTag::Tagged(_) => {
+                // Search the cache for the tag we're looking up
+                if let Some(cache_idx) = self.cache.tags.iter().position(|t| *t == tag) {
```

Comment:
So we do a linear search of the cache here?
I am quite surprised that that is beneficial...^^

Reply:
If the cache is small, this is a common cache optimization